### ISSUE TITLE
[CI][C++] Fail to install libboost-headers with mamba on AppVeyor

### ISSUE THREAD

**Issue by kou** [apache/arrow#41964]
### Describe the bug, including details regarding any error messages, version, and platform.

https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/49951613#L647

```text
error    libmamba Failed to download package from https://conda.anaconda.org/conda-forge/libboost-headers-1.85.0-h57928b3_1.conda (status 404)
Multi-download failed. Reason: Transfer finalized, status: 404 [https://conda.anaconda.org/conda-forge/libboost-headers-1.85.0-h57928b3_1.conda] 580 bytes
# >>>>>>>>>>>>>>>>>>>>>> ERROR REPORT <<<<<<<<<<<<<<<<<<<<<<
    Traceback (most recent call last):
      File "C:\Miniconda38-x64\lib\site-packages\conda\exception_handler.py", line 18, in __call__
        return func(*args, **kwargs)
      File "C:\Miniconda38-x64\lib\site-packages\mamba\mamba.py", line 959, in exception_converter
        raise e
      File "C:\Miniconda38-x64\lib\site-packages\mamba\mamba.py", line 952, in exception_converter
        exit_code = _wrapped_main(*args, **kwargs)
      File "C:\Miniconda38-x64\lib\site-packages\mamba\mamba.py", line 898, in _wrapped_main
        result = do_call(parsed_args, p)
      File "C:\Miniconda38-x64\lib\site-packages\mamba\mamba.py", line 767, in do_call
        exit_code = create(args, parser)
      File "C:\Miniconda38-x64\lib\site-packages\mamba\mamba.py", line 602, in create
        return install(args, parser, "create")
      File "C:\Miniconda38-x64\lib\site-packages\mamba\mamba.py", line 558, in install
        transaction.fetch_extract_packages()
    RuntimeError: Multi-download failed. Reason: Transfer finalized, status: 404 [https://conda.anaconda.org/conda-forge/libboost-headers-1.85.0-h57928b3_1.conda] 580 bytes
```

### Component(s)

C++, Continuous Integration

**Comment by h-vetinari** [comment]
These sort of download errors are hard to predict or reproduce (absent a big infrastructure outage). A simple restart is all that should be needed.

To make sure, I just installed that exact package as a test, and it worked without issue (to rule out even the _very_ rare cases of a package being not available in the CDN).

TL;DR: just restart such failures

**Comment by kou** [comment]
Hmm... These download errors persistent in our CI: https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/history

For example:

https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/49952303#L650

```text
error    libmamba Failed to download package from https://conda.anaconda.org/conda-forge/libexpat-2.6.2-h63175ca_0.conda (status 404)
Multi-download failed. Reason: Transfer finalized, status: 404 [https://conda.anaconda.org/conda-forge/libexpat-2.6.2-h63175ca_0.conda] 571 bytes
```

https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/49952259#L649

```text
error    libmamba Failed to download package from https://conda.anaconda.org/conda-forge/llvm-meta-5.0.0-0.tar.bz2 (status 404)
Multi-download failed. Reason: Transfer finalized, status: 404 [https://conda.anaconda.org/conda-forge/llvm-meta-5.0.0-0.tar.bz2] 565 bytes
```

**Comment by kou** [comment]
The URLs used by libmamba are 404 on my local environment:

```console
$ LANG=C wget https://conda.anaconda.org/conda-forge/libboost-headers-1.85.0-h57928b3_1.conda
--2024-06-05 06:15:17--  https://conda.anaconda.org/conda-forge/libboost-headers-1.85.0-h57928b3_1.conda
Resolving conda.anaconda.org (conda.anaconda.org)... 2606:4700::6813:9125, 2606:4700::6813:9025, 104.19.145.37, ...
Connecting to conda.anaconda.org (conda.anaconda.org)|2606:4700::6813:9125|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2024-06-05 06:15:18 ERROR 404: Not Found.
$ LANG=C wget https://conda.anaconda.org/conda-forge/libexpat-2.6.2-h63175ca_0.conda
--2024-06-05 06:15:34--  https://conda.anaconda.org/conda-forge/libexpat-2.6.2-h63175ca_0.conda
Resolving conda.anaconda.org (conda.anaconda.org)... 2606:4700::6813:9025, 2606:4700::6813:9125, 104.19.145.37, ...
Connecting to conda.anaconda.org (conda.anaconda.org)|2606:4700::6813:9025|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2024-06-05 06:15:35 ERROR 404: Not Found.
$ LANG=C wget https://conda.anaconda.org/conda-forge/llvm-meta-5.0.0-0.tar.bz2
--2024-06-05 06:15:47--  https://conda.anaconda.org/conda-forge/llvm-meta-5.0.0-0.tar.bz2
Resolving conda.anaconda.org (conda.anaconda.org)... 2606:4700::6813:9125, 2606:4700::6813:9025, 104.19.144.37, ...
Connecting to conda.anaconda.org (conda.anaconda.org)|2606:4700::6813:9125|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2024-06-05 06:15:47 ERROR 404: Not Found.
```

**Comment by kou** [comment]
In general, is URL pattern (`https://conda.anaconda.org/conda-forge/${PACKAGE}-${PACKAGE_VERSION}.{conda,tar.bz2}`) used by mamba correct?

**Comment by h-vetinari** [comment]
> In general, is URL pattern (`https://conda.anaconda.org/conda-forge/${PACKAGE}-${PACKAGE_VERSION}.{conda,tar.bz2}`) used by mamba correct?

I usually don't dig into this manually, so I'm not aware how it's resolved behind the scenes. Looking at the [files](https://anaconda.org/conda-forge/libboost-headers/files) & where they link, I see
```
https://anaconda.org/conda-forge/libboost-headers/1.85.0/download/win-64/libboost-headers-1.85.0-h57928b3_1.conda
```
however, that still does some link forwarding (i.e. download with `curl -L` is correct, but incorrect without `-L`). I also don't see a `Location:` under `curl -ILSs <url>`, so it's not obvious to me where it points to.

In general though, don't mess with manual URLs - conda should get the right information from the channeldata. I don't know if your appveyor instance just happens to have a very unstable network? Do you see these failures on any other CI provider?

I haven't ever seen these errors being repeatable. If that turns out to be the case, please raise an issue under https://github.com/conda/infrastructure/

**Comment by kou** [comment]
I haven't seen these errors in other CI jobs but they use cached Docker image.
I removed the cache and restarted a CI job that uses mamba:

https://github.com/apache/arrow/actions/runs/9375721796/job/25827082290#step:6:866

These errors aren't happen...

This may be related to Windows or AppVeyor. I'll debug on AppVeyor to collect information to report before I raise an issue somewhere.

**Comment by kou** [comment]
https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/49957015#L368

Using `conda` instead of `mamba` doesn't cause this problem. mamba related problem...?

**Comment by kou** [comment]
`mamba clean --all -y` before `mamba create ...` solves this: https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/49963459

AppVeyor may have invalid cache for mamba. Should we clean caches explicitly? Or should invalid cache be cleaned automatically?

**Comment by h-vetinari** [comment]
> AppVeyor may have invalid cache for mamba. Should we clean caches explicitly? Or should invalid cache be cleaned automatically?

Cleaning the cache sounds like a good idea IMO. Downloading fresh channel metadata is a good idea in any case, and especially if there's a stale cache in the image somehow.

**Comment by kou** [comment]
OK. Let's clean caches explicitly.

**Comment by pitrou** [comment]
Issue resolved by pull request 41977
https://github.com/apache/arrow/pull/41977
